Group-01/main.cpp

```
@@ -0,0 +1,1250 @@
+      var mode = document.getElementById("traversalMode").value;
+      if (mode == "DFS")
+        return 0;
+      if (mode == "BFS")
+        return 1;
+      if (mode == "A*")
+        return 2;
+      return 0;
+    });
+    if (mode == 2 && !aStarDestVertex.has_value()) {
+      // Show alert if trying to start an Astar traversal without a destination
+      EM_ASM({ alert("You cannot start an Astar traversal without selecting a destination."); });
+      return;
+    }
+
+    if (g.HasCycle()) {
+      EM_ASM({ alert("Can't traverse a Graph with a cycle"); });
+      return;
+    }
+
+    if (!traversal.has_value()) {
+      StartTraversal();
+      std::cout << "Started traversal" << std::endl;
+    };
+
+    std::cout << traversal.value().GetCurrentVertex() << std::endl;
+    if (traversal && traversal->AdvanceToNextNeighbor()) {
+      RedrawCanvas();
+    } else {
+      RedrawCanvas();
+      EM_ASM({
+        setTimeout(function() { alert("Traversal Finished!"); },
+                   150); // wait 150ms to let canvas update
+      });
+    }
+  }
+
+  /**
+   * Traverse all the way through a graph in one button press
+   */
+  void FullTraversal() {
+    if (!startVertex.has_value()) {
+      EM_ASM({ alert("Please select a starting vertex before starting the traversal."); });
+      return;
+    }
+
+    if (!traversal.has_value())
+      StartTraversal();
+
+    // Kick off the async step loop
+    emscripten_async_call(StepTraversalAsync, this,
+                          0); // 0ms to start immediately
+  }
+
+  // Export the graph to JSON
+  char *ExportGraph() {
+    std::string jsonStr = graphJson.ExportToJson();
+    char *result = (char *)malloc(jsonStr.length() + 1);
+    strcpy(result, jsonStr.c_str());
+    return result;
+  }
+
+  // Import the graph from JSON
+  bool ImportGraph(const char *jsonStr) {
+    bool success = graphJson.ImportFromJson(jsonStr);
+    if (success) {
+      RedrawCanvas();
+    }
+    return success;
+  }
```

I feel that silently failing on bad JSON can confuse users - adding an EM_ASM({ alert("Import failed") }) or a console log on failure would make debugging much easier. It will help if it gives immediate feedback rather than doing nothing.

---

Done